enhancements/automatic-catalog-switching.md

```diff
@@ -0,0 +1,251 @@
+third party catalogs. This proposal describes a simple approach to 
+switching the catalog image based on platform updates.
+
+## Motivation
+
+All third party Operators have a similar requirement to automatically upgrade the Operator if 
+the underlying platform gets upgraded, to ensure users are left with a compatible and
+supported Operator install.
+
+### Goals
+
+- Enable automatic changing of the Operator catalog image as part of a platform upgrade in 
+order to avoid the problem where cluster upgrades leave Operator installations in 
+unsupported states or without a continued update path.
+- Use a simple URL templating scheme to achieve 80% of the use use case seen today
+
+### Non-Goals
+
+- Using advanced solutions
+  - a catalog of catalogs via declarative config that automatically determine which is the correct catalog
+  - new APIs that stamp out CatalogSource objects
+  - proxies at the registry side serving the catalog images
+
+## Proposal
+
+This enhancement should enable catalog image references to contain template values that can be replaced with
+actual values from the underlying cluster. These templates should allow for flexibility so that an
+image reference can be be composed in such a way that allows third parties to create image name / tag 
+in whatever way they choose. These well known template names will be replaced by a lookup 
+mechanism to discover its corresponding value. The current proposal concentrates on 
+template names that correspond to portions of a version (following semver major, minor, and patch). 
+There are other possible template values provided but version is the primary focus.
+Note that all templates must be surrounded by `{` and `}`, contain no spaces, and be all lower case.
+
+### required templates
+
+- {kube_major_version}
+  - example resolved value: 1
+- {kube_minor_version}
+  - example resolved value: 19
+- {kube_patch_version}
+  - example resolved value: 0
+- {olm_major_version}
+  - example resolved value: 0
+- {olm_minor_version}
+  - example resolved value: 18
+- {olm_patch_version}
+  - example resolved value: 1
+- {platform_architecture}
+  - example resolved value: x86_64
+
+### suggested downstream templates (only available in downstream versions of OLM)
+
+- {ocp_major_version}
+  - example resolved value: 1
+- {ocp_minor_version}
+  - example resolved value: 1
+- {ocp_patch_version}
+  - example resolved value: 1
+- {ocp_update_channels}
+  - example resolved value: fast
+
+### Upstream OLM
+
+Consider the following example for upstream OLM, where a catalog can be switched out based on the Kubernetes version in use:
+
+```yaml
+apiVersion: operators.coreos.com/v1alpha1
+kind: CatalogSource
+metadata:
+  name: dynamic-catalog
+  namespace: olm
+spec:
+  sourceType: grpc
+  image: quay.io/kube-release-v{kube_major_version}/catalog:v{kube_major_version}.{kube_minor_version}
+  displayName: My always up2date Catalog
+  publisher: Red Hat
```

Comment:
We talked about this in the community call, but one suggestion that sounded nice to me was to have this as an annotation:

```yaml
apiVersion: operators.coreos.com/v1alpha1
kind: CatalogSource
metadata:
  name: dynamic-catalog
  namespace: olm
  annotations:
    olm.imageTemplate: "quay.io/kube-release-v{kube_major_version}/catalog:v{kube_major_version}.{kube_minor_version}"
spec:
  sourceType: grpc
  image: quay.io/kube-release-v1/catalog:v1.20
  displayName: My always up2date Catalog
  publisher: Red Hat
```

Then we would write a controller that:
- watches kube versions
- watches CatalogSources that have the `olm.imageTemplate` annotation
- when the resolved values of the imageTemplate doesn't match `spec.image`, it updates `spec.image`

This could be shipped as an optional operator packaged for olm, or it could just be a separate controller within the catalog-operator. I like the idea of it being separate, because it would make it easier to add support for other distributions.

Reply:
In case you did not see it, I opened https://github.com/operator-framework/enhancements/pull/87 to cover this idea.